### PR TITLE
screenshot and remappable keys function

### PR DESCRIPTION
**Pull request by jopadan**
- switch from KeyboardState to InputState allowing other input device type states
- use char32_t instead of unsigned int for KeyCode
- add remappable keys present in original portable release like center_view and weapon_next/prev
- allow mouse control remap
- allow rotate/wrap for msaa and pixel_size
- fix character input in window resolution video menu
- add file path string functions for C++ < 17
- modify WriteTGA to support RGBA32 flipped input if palette equals nullptr
- add SystemEvent support to KeyChecker
- add screenshot implementation #49
- add LICENSE
- add add-on list in README.md

**TODO:**
- KeyChecker InputState mouse wheel support (only SystemEvent works atm)
- multiple exclusion key event map
- allow list of input device types to be mapped

### CODE REVIEW

File: src/system_window.cpp [hunk 1]
```diff
@@ -290,7 +295,7 @@ SystemWindow::SystemWindow( Settings& settings )
 			"PanzerChasm",
 			SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED,
 			width, height,
-			( (is_opengl || use_gl_context_for_software_renderer_) ? SDL_WINDOW_OPENGL : 0 ) | ( fullscreen ? SDL_WINDOW_FULLSCREEN : 0 ) | SDL_WINDOW_SHOWN );
+			( (is_opengl || use_gl_context_for_software_renderer_) ? SDL_WINDOW_OPENGL : 0 ) | ( fullscreen ? SDL_WINDOW_FULLSCREEN : 0 ) | SDL_WINDOW_SHOWN  | SDL_WINDOW_KEYBOARD_GRABBED);
```
Comment (Panzerschrek) [inline src/system_window.cpp]:
```
error: use of undeclared identifier 'SDL_WINDOW_KEYBOARD_GRABBED'; did you mean 'SDL_WINDOW_INPUT_GRABBED'
```

File: src/menu.cpp [hunk 2]
```diff
@@ -1021,15 +1036,39 @@ class ControlsMenu final : public MenuBase
...
+	{ "Permanent MLook"     , SettingsKeys::key_perm_mlook    , KeyCode::Enter  },
+	{ "Temporary MLook"     , SettingsKeys::key_temp_mlook    , KeyCode::BackSlash },
+	{ "Strafe On"           , SettingsKeys::key_strafe_on     , KeyCode::Mouse4 },
+	{ "Look Up"		, SettingsKeys::key_look_up	  , KeyCode::Up },
+	{ "Look Down"		, SettingsKeys::key_look_down	  , KeyCode::Down },
+	{ "Turn Left"		, SettingsKeys::key_turn_left	  , KeyCode::Left },
+	{ "Turn Right"		, SettingsKeys::key_turn_right	  , KeyCode::Right },
+	{ "Center View"         , SettingsKeys::key_center_view   , KeyCode::C },
+	*/
 };
```
Comment (Panzerschrek) [inline src/menu.cpp]: no_place_for_new_menu_items
This looks not so good.

Comment (Panzerschrek) [reply src/menu.cpp]: I think it is important to keep old menu style and avoid such huge redesign. Maybe add new settings into another submenu (like _Advanced Controls_)?

File: src/system_event.hpp [hunk 3]
```diff
@@ -116,7 +134,7 @@ struct SystemEvent
 typedef std::vector<SystemEvent> SystemEvents;
 
 typedef std::bitset< static_cast<unsigned int>( SystemEvent::KeyEvent::KeyCode::KeyCount ) > KeyboardState;
-typedef std::bitset< static_cast<unsigned int>( SystemEvent::MouseKeyEvent::Button::ButtonCount ) > MouseState;
+typedef std::bitset< static_cast<unsigned int>( SystemEvent::MouseKeyEvent::Button::ButtonCount) + 2  > MouseState;
```
Comment (Panzerschrek) [inline src/system_event.hpp]: Avoid using magic constant _2_. Give it a name.

File: src/system_window.cpp [hunk 4]
```diff
@@ -75,8 +81,7 @@ static SystemEvent::MouseKeyEvent::Button TranslateMouseButton( const Uint8 butt
 	case SDL_BUTTON_MIDDLE: return Button::Middle;
 	};
 
-	return Button::Unknown;
-}
+	return Button::Unknown;	}
```
Comment (Panzerschrek) [inline src/system_window.cpp]: Avoid unnecessary formatting changes.

File: default_PanzerChasm.cfg [hunk 5]
```diff
@@ -1,48 +1,61 @@
...
-"r_software_gl_update_smooth" "0"
-"r_software_rendering" "1"
-"r_software_scale" "1"
-"r_software_use_gl_screen_update" "0"
-"r_window_height" "600"
-"r_window_width" "800"
-"s_nosound" "0"
-"s_volume" "0.500000"
-"sv_always_run" "1"
+"cd_volume" "0.500000"
```
Comment (Panzerschrek) [inline default_PanzerChasm.cfg]: Let's avoid making unnecessary changes in this file. It would be better to add default values for new settings manually, rather just replacing the config file with newly-generated.

File: src/host.cpp [hunk 6]
```diff
@@ -244,7 +244,7 @@ bool Host::Loop()
 		if( console_ != nullptr )
 			console_->Draw();
 
-		if( settings_.GetOrSetBool( "cl_draw_fps", false ) )
+		if( settings_.GetOrSetBool( "cl_draw_fps", true ) )
```
Comment (Panzerschrek) [inline src/host.cpp]: Let's avoid changing this. I think for most players it's not important to show FPS counter.

File: src/menu.cpp [hunk 7]
```diff
@@ -1519,11 +1560,18 @@ MenuBase* GraphicsMenu::Select()
 		PlayMenuSound( Sound::SoundId::MenuChange );
 	}
 
+	int pixel_size;
+	unsigned int msaa_level;
```
Comment (Panzerschrek) [inline src/menu.cpp]: Avoid uninitialized variables. Give some safe initial value (0?).

File: src/menu.cpp (hunk 2 above)
Comment (Panzerschrek) [reply src/menu.cpp]: Also it is important to support screen resolutions as small as 320x200 (see `GameConstants::min_screen_width`).

**Review by Panzerschrek — CHANGES_REQUESTED** [review]

File: src/system_window.cpp (hunk 1 above)
Comment (Panzerschrek) [reply src/system_window.cpp]: I still use Ubuntu 18.04 with SDL 2.0.8, because of that we still need to support such old SDL2 version.

File: src/menu.cpp (hunk 2 above)
Comment (jopadan) [reply src/menu.cpp]: Shortened the strings to draw menu at resolution 640x480 resolving the overdraw issue.
Fixed keys using updated KeyChecker supporting SystemEvent too

File: src/system_window.cpp (hunk 1 above)
Comment (jopadan) [reply src/system_window.cpp]: Switched to SDL_SetWindowKeyboardGrab

File: CMakeLists.txt [hunk 8]
```diff
@@ -2,7 +2,7 @@ cmake_minimum_required(VERSION 3.1)
 
 project(Chasm-Reverse)
 
-set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD 17)
```
Comment (Panzerschrek) [inline CMakeLists.txt]: Is it really necessary to use newer C++ standard?
If no - let's preserve standard 11.

File: src/system_event.hpp [hunk 9]
```diff
@@ -127,4 +149,58 @@ struct InputState
 const char* GetKeyName( SystemEvent::KeyEvent::KeyCode key_code );
 bool KeyCanBeUsedForControl( SystemEvent::KeyEvent::KeyCode key_code );
 
+using KeyCode= SystemEvent::KeyEvent::KeyCode;
+
+class KeyChecker
```
Comment (Panzerschrek) [inline src/system_event.hpp]: Let's move this class into separate pair of hpp/cpp files.

File: src/shared_settings_keys.hpp [hunk 10]
```diff
@@ -19,15 +19,50 @@ const char cd_volume[]= "cd_volume";
...
+const char key_small_hud_on[]= "in_key_small_hud_on";
+const char key_weapon_change[]= "in_key_weapon_change";
+const char key_weapon_next[]= "in_key_weapon_next";
+const char key_weapon_prev[]= "in_key_weapon_prev";
+const char key_menu[]= "in_key_menu";
+const char key_menu_mouse[]= "in_key_menu_mouse";
+const char key_menu_left[]= "in_key_menu_left";
+const char key_menu_right[]= "in_key_menu_right";
+/* addiotional optional keys unnecessary with mouse */
+const char key_inventory_use[]= "in_key_inventory_change";
```
Comment (Panzerschrek) [inline src/shared_settings_keys.hpp]: Invertory? _Chasm: the Rift_ has an inventory, like _Heretic_/_Hexen_?
Let's remove unnecessary controls.

File: src/system_window.cpp [hunk 11]
```diff
@@ -291,10 +297,11 @@ SystemWindow::SystemWindow( Settings& settings )
 			SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED,
 			width, height,
 			( (is_opengl || use_gl_context_for_software_renderer_) ? SDL_WINDOW_OPENGL : 0 ) | ( fullscreen ? SDL_WINDOW_FULLSCREEN : 0 ) | SDL_WINDOW_SHOWN );
-
 	if( window_ == nullptr )
 		Log::FatalError( "Can not create window" );
 
+	SDL_SetWindowKeyboardGrab(window_, SDL_TRUE);
```
Comment (Panzerschrek) [inline src/system_window.cpp]:
```
error: use of undeclared identifier 'SDL_SetWindowKeyboardGrab'
        SDL_SetWindowKeyboardGrab(window_, SDL_TRUE);
        ^
```
This doesn't exist in SDL 2.0.9.

File: src/menu.cpp [hunk 12]
```diff
@@ -1021,15 +1035,39 @@ class ControlsMenu final : public MenuBase
...
-	{ "Move Forward"	, SettingsKeys::key_forward	, KeyCode::W },
-	{ "Move Backward"	, SettingsKeys::key_backward	, KeyCode::S },
-	{ "Strafe Left"		, SettingsKeys::key_step_left	, KeyCode::A },
-	{ "Strafe Right"	, SettingsKeys::key_step_right	, KeyCode::D },
-	{ "Turn Left"		, SettingsKeys::key_turn_left	, KeyCode::Left },
-	{ "Turn Right"		, SettingsKeys::key_turn_right	, KeyCode::Right },
-	{ "Look Up"		, SettingsKeys::key_look_up	, KeyCode::Up },
-	{ "Look Down"		, SettingsKeys::key_look_down	, KeyCode::Down },
-	{ "Jump"		, SettingsKeys::key_jump	, KeyCode::Space },
+	/* essential controls */
```
Comment (Panzerschrek) [inline src/menu.cpp]: controls_new
controls_old

Looks better, but still - not good enough.
I still suggest to add controls submenu with advanced controls - all added in this PR and (maybe) also controls for weapons selection.

**Review by Panzerschrek — CHANGES_REQUESTED** [review]

File: src/menu.cpp [hunk 13]
```diff
@@ -1021,15 +1035,39 @@ class ControlsMenu final : public MenuBase
...
+	{ "Speed Up"            , SettingsKeys::key_speed_up      , KeyCode::LeftShift },
+	{ "Always Run"          , SettingsKeys::key_always_run    , KeyCode::CapsLock },
+	{ "Jump"		, SettingsKeys::key_jump	  , KeyCode::Mouse2 },
+	{ "Fire"                , SettingsKeys::key_fire	  , KeyCode::Mouse1 },
+	/* original game and no mouse wheel user backward compatibility */
+	{ "Change Weapon"	, SettingsKeys::key_weapon_change , KeyCode::Mouse3 },
+	/* additional controls introduced in PanzerChasm with mouse wheel support */
+	{ "Next Weapon"         , SettingsKeys::key_weapon_next   , KeyCode::MouseWheelDown },
+	{ "Prev Weapon"     , SettingsKeys::key_weapon_prev   , KeyCode::MouseWheelUp },
+	{ "Menu"                , SettingsKeys::key_menu          , KeyCode::Escape },
```
Comment (Panzerschrek) [inline src/menu.cpp]: I think menu key (and also console key) should not be configurable. Who really needs this?

**Comment by Panzerschrek** [comment]
@jopadan 
Generally i suggest to open an issue first prior to making PRs in order to discuss what is right way to do something prior to actually doing it.

File: src/shared_settings_keys.hpp (hunk 10 above)
Comment (Panzerschrek) [reply src/shared_settings_keys.hpp]: I am generally against adding additional functionality, that doesn't exist in original game. Even minor changes like sprint or zoom can break original game experience.

File: src/system_window.cpp (hunk 11 above)
Comment (jopadan) [reply src/system_window.cpp]: Removed it for now, but CapsLock doesn't work without it.

File: src/menu.cpp (hunk 12 above)
Comment (jopadan) [reply src/menu.cpp]: screenshot
Adapted control menu style to look like the original portable 1.05 version distributed with GOG/Steam CHASMPR.CUE/BIN plus added weapon next/prev and added keypad scancodes.

**Comment by Panzerschrek** [comment]
This PR now contains too much changes in order to accept it.
@jopadan , split it into several small logical parts (create several different PRs) or close it at all.
